File: src/api/registration.rs
//! Registration Memory: the ten buttons, banks, Memorize, Freeze and the Registration
//! Sequence (docs/registration.md).

use crate::registration::{Group, Groups, SequenceEnd};
use serde::{Deserialize, Serialize};

/// Button indices are 0-based (0-9 = the panel's [1]-[10]).
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "camelCase", rename_all_fields = "camelCase")]
pub enum RegistrationCmd {
    /// A REGISTRATION MEMORY button as the panel has it: recalls the button, or, while
    /// MEMORY is armed (`toggleRegistMemory`), memorizes the panel into it.
    PressRegist { index: u8 },
    /// Recall a button (refused if it is empty).
    RecallRegist { index: u8 },
    /// Memorize the panel into a button (the `memorizeGroups`), replacing what it held.
    MemorizeRegist { index: u8 },
    /// The MEMORY button: arm (or disarm) Memorize for the next button press.
    ToggleRegistMemory,
    /// Tick or untick a group in the Memory window (what Memorize stores).
    SetMemorizeGroup { group: Group, on: bool },
    /// Empty a button (Regist Bank Edit, Delete).
    ClearRegist { index: u8 },
    /// Rename a button (Regist Bank Edit, Rename).
    RenameRegist { index: u8, name: String },
    /// REGIST BANK -/+: the previous/next bank file in the folder.
    StepRegistBank { delta: i8 },
    /// Load a bank file (its path, as `registration.banks` lists it).
    SelectRegistBank { path: String },
    /// Start a new, empty, unsaved bank.
    NewRegistBank,
    /// Save the bank: to its file, or with `name` as a new file in the folder (Save As).
    /// Saving with the name of another bank's file is refused unless `overwrite` is set.
    SaveRegistBank {
        name: Option<String>,
        #[serde(default, skip_serializing_if = "std::ops::Not::not")]
        overwrite: bool,
    },
    /// Registration Freeze on/off (the FREEZE button).
    SetFreeze { on: bool },
    ToggleFreeze,
    /// Tick or untick a group on the Regist Freeze display.
    SetFreezeGroup { group: Group, on: bool },
    /// Program the bank's Registration Sequence: buttons (0-9) in order, and the end action.
    SetRegistSequence { steps: Vec<u8>, end: SequenceEnd },
    /// Registration Sequence on/off.
    SetRegistSequenceOn { on: bool },
    ToggleRegistSequence,
    /// Regist +/- (`delta` 1 / -1): the next/previous step of the sequence.
    StepRegistSequence { delta: i8 },
}

/// Registration Memory.
#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RegistrationState {
    /// The bank in use.
    pub bank: BankState,
    /// The bank files in the folder, in order (REGIST BANK -/+ step through them).
    pub banks: Vec<BankFile>,
    /// The folder bank files are saved to (None: saving is off, e.g. an offline session).
    pub folder: Option<String>,
    /// Always 10: buttons [1]-[10] (Regist Bank Info).
    pub buttons: Vec<RegistButton>,
    /// The button last recalled or memorized (lit red), 0-based.
    pub selected: Option<u8>,
    /// MEMORY is armed: the next button press memorizes.
    pub memory: bool,
    /// The Memory window's ticked groups.
    pub memorize_groups: Groups,
    /// Registration Freeze is on.
    pub freeze: bool,
    /// The Regist Freeze display's ticked groups (they stay unchanged on recall while
    /// `freeze` is on).
    pub freeze_groups: Groups,
    pub sequence: SequenceState,
    /// A recall is waiting for the style it loads to take over (the bar line when playing);
    /// the rest of the registration follows then.
    pub pending: bool,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BankState {
    pub name: String,
    /// Its file (None: a new bank not saved yet).
    pub path: Option<String>,
    /// Changed since it was loaded or saved.
    pub dirty: bool,
    /// Its place in `banks` (None when unsaved or outside the folder).
    pub position: Option<usize>,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BankFile {
    pub name: String,
    pub path: String,
}

/// One Registration Memory button, as Regist Bank Info shows it.
#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RegistButton {
    /// 0-based.
    pub index: u8,
    /// It holds a registration (lit blue, or red when selected).
    pub stored: bool,
    pub name: String,
    /// The groups it memorized.
    pub groups: Groups,
    /// The style it loads (name), if it stores one.
    pub style: Option<String>,
    pub tempo: Option<f64>,
    /// Right 1, Right 2, Right 3, Left: the voice names it sets, if it stores them.
    pub voices: Vec<RegistVoice>,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RegistVoice {
    pub name: String,
    pub on: bool,
}

#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SequenceState {
    pub on: bool,
    /// Buttons (0-based) in order.
    pub steps: Vec<u8>,
    pub end: SequenceEnd,
    /// The step last recalled (0-based into `steps`); None before the first.
    pub position: Option<usize>,
}

File: app/src/lib/api/registration.ts
// Registration Memory and the Playlist (docs/app-api.md "Registration Memory",
// "Playlist"; docs/registration.md). Mirrors src/api/registration.rs and
// src/api/playlist.rs field for field.

/** A Genos Freeze / Memorize group. */
export type RegistGroup =
  | 'style' | 'voice' | 'harmonyArp' | 'multiPad' | 'tempo' | 'transpose' | 'chordLooper' | 'liveControl' | 'assignable'

/** What the Registration Sequence does past its last step. */
export type SequenceEnd = 'stop' | 'top' | 'next'

export type PlaylistSort = 'normal' | 'aToZ' | 'zToA'

/** A Playlist record: a bank file (and the button to recall after loading it), or a style. */
export type PlaylistRecord =
  | { name: string; kind: 'bank'; path: string; regist?: number | null }
  | { name: string; kind: 'style'; path: string }

/** Registration Memory commands. Buttons are 0-based (0–9 = the panel's [1]–[10]). */
export type RegistrationCmd =
  /** A REGISTRATION MEMORY button: recall, or memorize while MEMORY is armed. */
  | { type: 'pressRegist'; index: number }
  | { type: 'recallRegist'; index: number }
  | { type: 'memorizeRegist'; index: number }
  /** The MEMORY button: the next button press memorizes. */
  | { type: 'toggleRegistMemory' }
  | { type: 'setMemorizeGroup'; group: RegistGroup; on: boolean }
  | { type: 'clearRegist'; index: number }
  | { type: 'renameRegist'; index: number; name: string }
  /** REGIST BANK −/+. */
  | { type: 'stepRegistBank'; delta: number }
  | { type: 'selectRegistBank'; path: string }
  | { type: 'newRegistBank' }
  /** Save to its file (`name` null) or as a file of that name in the folder; another bank's file of that name is only replaced with `overwrite`. */
  | { type: 'saveRegistBank'; name: string | null; overwrite?: boolean }
  | { type: 'setFreeze'; on: boolean }
  | { type: 'toggleFreeze' }
  | { type: 'setFreezeGroup'; group: RegistGroup; on: boolean }
  | { type: 'setRegistSequence'; steps: number[]; end: SequenceEnd }
  | { type: 'setRegistSequenceOn'; on: boolean }
  | { type: 'toggleRegistSequence' }
  /** Regist +/−: the sequence's next/previous step. */
  | { type: 'stepRegistSequence'; delta: number }
  /** Regist +/− from a pedal: the sequence while it is on and programmed, else the bank's
   * next/previous stored button (it stops at either end). */
  | { type: 'stepRegist'; delta: number }

/** Playlist commands. Record `index` is the position in the file (`PlaylistRow.index`). */
export type PlaylistCmd =
  | { type: 'newPlaylist' }
  | { type: 'loadPlaylist'; path: string }
  | { type: 'savePlaylist'; name: string | null; overwrite?: boolean }
  | { type: 'addPlaylistRecord'; record: PlaylistRecord }
  | { type: 'addCurrentBank' }
  | { type: 'addCurrentStyle' }
  | { type: 'appendPlaylist'; path: string }
  | { type: 'setPlaylistRecord'; index: number; record: PlaylistRecord }
  | { type: 'movePlaylistRecord'; index: number; delta: number }
  | { type: 'deletePlaylistRecord'; index: number }
  | { type: 'setPlaylistSort'; sort: PlaylistSort }
  | { type: 'loadPlaylistRecord'; index: number }
  | { type: 'stepPlaylist'; delta: number }

export interface RegistButton {
  index: number
  /** It holds a registration (lamp blue, or red when selected). */
  stored: boolean
  name: string
  /** The groups it memorized. */
  groups: RegistGroup[]
  style: string | null
  tempo: number | null
  /** Right 1, Right 2, Right 3, Left; empty when it stores no parts. */
  voices: { name: string; on: boolean }[]
}

export interface RegistrationState {
  bank: { name: string; path: string | null; dirty: boolean; position: number | null }
  banks: { name: string; path: string }[]
  /** Where banks are saved; null when saving is off. */
  folder: string | null
  /** Always 10. */
  buttons: RegistButton[]
  /** The button last recalled or memorized. */
  selected: number | null
  /** MEMORY is armed. */
  memory: boolean
  memorizeGroups: RegistGroup[]
  freeze: boolean
  freezeGroups: RegistGroup[]
  sequence: { on: boolean; steps: number[]; end: SequenceEnd; position: number | null }
  /** A recall waits for its style to take over (the bar line). */
  pending: boolean
}

export interface PlaylistRow {
  /** Its position in the file (what the commands take). */
  index: number
  record: PlaylistRecord
  /** Its bank or style file is not there. */
  missing: boolean
}

export interface PlaylistState {
  name: string
  path: string | null
  dirty: boolean
  sort: PlaylistSort
  /** In display order. */
  records: PlaylistRow[]
  /** The record last loaded (file position). */
  current: number | null
  playlists: { name: string; path: string }[]
  folder: string | null
}

/** The file name (less its extension) a bank or playlist called `name` is saved under, as
 * the backend makes it (`registration::file_name`): `/`, `\`, `:` and NUL become `_`,
 * leading dots go, and an empty name is Untitled. */
export function fileStem(name: string): string {
  // eslint-disable-next-line no-control-regex
  const clean = name.trim().replace(/[/\\:\u0000]/g, '_').replace(/^\.+/, '')
  return clean || 'Untitled'
}

/** Names `a` and `b` save to the same file: the Mac's file system (APFS) ignores case. */
export const sameFile = (a: string, b: string) => fileStem(a).toLowerCase() === fileStem(b).toLowerCase()

/** The groups in the Genos's order, with the names the Memory and Freeze windows use. */
export const REGIST_GROUPS: { id: RegistGroup; name: string }[] = [
  { id: 'style', name: 'Style' },
  { id: 'voice', name: 'Voice' },
  { id: 'harmonyArp', name: 'Harmony/Arp' },
  { id: 'multiPad', name: 'Multi Pad' },
  { id: 'tempo', name: 'Tempo' },
  { id: 'transpose', name: 'Transpose' },
  { id: 'chordLooper', name: 'Chord Looper' },
  { id: 'liveControl', name: 'Live Control' },
  { id: 'assignable', name: 'Assignable' },
]

export const SEQUENCE_ENDS: { id: SequenceEnd; name: string }[] = [
  { id: 'stop', name: 'Stop' },
  { id: 'top', name: 'Top' },
  { id: 'next', name: 'Next bank' },
]

export function emptyRegistration(): RegistrationState {
  return {
    bank: { name: 'New Bank', path: null, dirty: false, position: null },
    banks: [],
    folder: null,
    buttons: Array.from({ length: 10 }, (_, index) => ({ index, stored: false, name: '', groups: [], style: null, tempo: null, voices: [] })),
    selected: null,
    memory: false,
    memorizeGroups: REGIST_GROUPS.map((g) => g.id),
    freeze: false,
    freezeGroups: [],
    sequence: { on: false, steps: [], end: 'stop', position: null },
    pending: false,
  }
}

export function emptyPlaylist(): PlaylistState {
  return { name: 'New Playlist', path: null, dirty: false, sort: 'normal', records: [], current: null, playlists: [], folder: null }
}
